distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso

```diff
@@ -0,0 +1,586 @@
+import project.Error.Error
+import project.Errors.Common.Arithmetic_Error
+import project.Errors.Illegal_Argument.Illegal_Argument
+import project.Nothing.Nothing
+import project.Panic.Panic
+import project.Warning.Warning
+from project.Data.Boolean import Boolean, False, True
+from project.Data.Numbers import Integer, Float, Number, Number_Parse_Error
+from project.Data.Ordering import Comparable, Ordering
+from project.Errors.Common import Loss_Of_Numeric_Precision
+
+polyglot java import java.lang.ArithmeticException
+polyglot java import java.lang.NumberFormatException
+polyglot java import java.math.BigDecimal
+polyglot java import org.enso.base.numeric.Conversion_Result
+polyglot java import org.enso.base.numeric.Decimal_Utils
+
+## Decimal is the type of decimal (base-10-scaled) nubmers in Enso. An Enso
+   `Decimal` is a wrapper around a Java `BigDecimal`.
+
+   `Decimal` values have more overhead than standard `Float` values, but they
+   have the advantage of being able to precisely specify values such as `0.1`,
+   which is important for certain kinds of calculations, such as currency.
+
+   A `Decimal` value is represented internally by a Java `BigInteger` "unscaled
+   value" and a "scale value". The numerical value of the `Decimal` is
+   `(unscaledValue * 10^(-scale))`. Scale values are maintained automatically by
+   the constructors and numerical operations.
+
+   Scale values can allow distinctions between values that would be identical as
+   `Float`s. For example, the following values have different internal
+   representations:
+
+       a = Decimal.new "2.0"
+       b = Decimal.new "2.00"
+       a == b
+       # => True
+
+   These two values have different internal representations, but they are still
+   considered the same value by `==`.
+
+   All of the constructors, and many of the operations, can take an additional
+   `Math_Context` value, which can specify two things:
+
+   - precision: the number of decimal digits to use to represent a value or
+     result; results are rounded to this precision.
+   - rounding_mode: the method to use for rounding. See `Rounding_Mode` for
+     details about each rounding method.
+
+   When a `Math_Context` value is used in an operation, and causes a loss of
+   precision, a `Loss_Of_Numeric_Precision` warning is attached to the result.
+
+   If no `Math_Context` value is supplied, the default is to make all operations
+   exact. A `Math_Context` with a precision of `0` has the same effect.
+
+   A `Decimal` can represent any `Float` precisely. However, some `Float`
+   literals, such as `0.1`, cannot be represented exactly to infinite
+   precision by a `Float`. For this reason, constructing a `Decimal` from
+   a `Float` always attaches a `Loss_Of_Numeric_Precision` warning to the
+   result. To avoid this problem, it is recommended to create `Decimal`s from
+   fractional values by passing the value in as a `Text`, where possible and
+   convenient.
```

Comment:
btw. for longer term, shall we introduce a big-decimal literal? Then we'd be able to avoid this issue mostly.

cc: @jdunkerley